EIPS/eip-5283.md

```diff
@@ -0,0 +1,107 @@
+
+### Alternatives
+
+There are alternative designs to implement RPGs on the EVM:
+
+1.	[EIP-1153: Transient storage opcodes](https://eips.ethereum.org/EIPS/eip-1153) or [`SSTORE_COUNT`](https://github.com/ethereum/EIPs/issues/119). The first does reduce the cost of an RPG, but does not address the problem of parallelization. The second enables execution parallelization but it is way more complex as it returns the number `SSTORE` opcodes executed, not the number of reentrant calls. Reentrancy must be deducted from this value. Also the `SSTORE` count is not currently tracked by EVM. Also, implementing new opcodes is discuraged if the same functionality can be implemented using precompiles.
+
+2. A new `LOCKCALL` opcode is introduced that works quite similar to `STATICALL` but only blocks storage writes in the caller contract. This results in cheaper RPG, but has two drawbacks: it doesn't allow some contract functions to be free of the RPG, and it intruduces a new opcode which requires modifying compilers, debuggers and static analysys tools.
+
+
+### Gas cost
+A gast cost of 100 represents a worst case resource consumption, which occurs when the stack is almost full (approximately 400 addresses) and it is fully scanned. As the stack is always present in RAM, the scanning is fast.  Note: Once code is implemented in geth, it can be benchmarked and the cost can be re-evaluated, as it may result to be lower in practice. As a precompile call currently costs 700 gas, the cost of stack scanning has a low impact on the total cost of the precompile call (800 gas in totall).
+
+The storage-based RPG currently costs 200 gas (because of the savings introduced in [EIP-1283](https://eips.ethereum.org/EIPS/eip-1283). Using the `Semaphore` precompile as a reentrancy check would currently costs 800 gas (a single call from one of the function modifiers). While this cost is higher than the traditional RPG cost, and therefore discourages its use,  it is still much lower than the pre-EIP1283 cost. If a reduction in precompile call cost (such as [EIP-1109](https://eips.ethereum.org/EIPS/eip-1109) or [EIP-2046](https://eips.ethereum.org/EIPS/eip-2046)) is implemented, then the cost of using the `Semaphore` precompile will be reduced to approximately 140 gas, below the current 200 gas consumed by a storage-based RPG. To encourage to use the precompile-based RPG, it is suggested that this EIP is implemented together with EIP-2046.
```

Comment:
```suggestion
The storage-based RPG currently costs 200 gas (because of the savings introduced in [EIP-1283](./eip-1283.md). Using the `Semaphore` precompile as a reentrancy check would currently cost 800 gas (a single call from one of the function modifiers). While this cost is higher than the traditional RPG cost and therefore discourages its use, it is still much lower than the pre-EIP-1283 cost. If a reduction in precompile call cost (such as [EIP-1109](./eip-1109.md) or [EIP-2046](./eip-2046.md)) is implemented, then the cost of using the `Semaphore` precompile will be reduced to approximately 140 gas, below the current 200 gas consumed by a storage-based RPG. To encourage to use of the precompile-based RPG, it is suggested that this EIP is implemented together with EIP-2046.
```